libtransmission/peer-io.cc

```diff
@@ -144,9 +144,9 @@ static void canReadWrapper(tr_peerIo* io_in)
     while (!done && !err)
     {
         size_t piece = 0;
-        size_t const old_len = io->readBufferSize();
-        int const ret = io->canRead(io.get(), io->userData, &piece);
-        size_t const used = old_len - io->readBufferSize();
+        auto const old_len = io->readBufferSize();
+        auto const read_state = io->canRead == nullptr ? READ_ERR : io->canRead(io.get(), io->userData, &piece);
```

Comment:
It's an improvement, but it's not threadsafe.
I would have copied io->canRead to a local variable before testing it for nullptr and using it.

Reply:
> It's an improvement, but it's not threadsafe. I would have copied io->canRead to a local variable before testing it for nullptr and using it.

I'm not sure I understand the thread safety question here -- where is `io->canRead` ever assigned or read from outside the session thread?

Reply:
If all is on the same thread, then look at line 133:
```
    if (io->canRead == nullptr)
    {
        return;
    }
```
Meaning that you're checking twice for the same condition.

Reply:
I suspect it's happening because of the `while()` loop that happens after the nullptr check. If the pointer gets cleared by some callback code invoked by canRead() then the callback can still be nullptr.